### ISSUE TITLE
FRP does not work with latest Chromium (likely due to changes in handshake)

### ISSUE THREAD

**Issue by ProfFan** [fatedier/frp#2500]
**[REQUIRED] hat version of frp are you using** 

Version: 0.37

**[REQUIRED] What operating system and processor architecture are you using**
OS: Ubuntu 20.04
CPU architecture: amd64

**[REQUIRED] description of errors**

"ERROR_CONNECTION_RESET" when trying to visit a website served using FRP with HTTPS (TLS).
Error log shows that the clientHello is sent and then the server returns TCP reset.

However, any other browser than Chromium (92.0.4515 currently) works fine.

**confile**

**log file**
```
Jul 29 23:50:00 us-atl frps[1384]: 2021/07/29 23:50:00 [I] [proxy.go:162] [] [] get a user connection [:48712]
Jul 29 23:50:00 us-atl frps[1384]: 2021/07/29 23:50:00 [D] [control.go:248] [] get work connection from pool
Jul 29 23:50:00 us-atl frps[1384]: 2021/07/29 23:50:00 [D] [proxy.go:103] [] [] get a new work connection: [:44742]
Jul 29 23:50:00 us-atl frps[1384]: 2021/07/29 23:50:00 [D] [proxy.go:273] [] [] join connections, workConn(l[:7000] r[:44742]) userConn(l[:443] r[:48712])
Jul 29 23:50:00 us-atl frps[1384]: 2021/07/29 23:50:00 [D] [control.go:219] [] new work connection registered
Jul 29 23:50:00 us-atl frps[1384]: 2021/07/29 23:50:00 [D] [proxy.go:283] [] [bitwarden] join connections closed
Jul 29 23:50:03 us-atl frps[1384]: 2021/07/29 23:50:03 [D] [vhost.go:147] get hostname from http/https request error: readHandshake: extensionsLen[1375] is not equal to dataLen[910]
Jul 29 23:50:03 us-atl frps[1384]: 2021/07/29 23:50:03 [D] [vhost.go:147] get hostname from http/https request error: readHandshake: extensionsLen[1377] is not equal to dataLen[908]
Jul 29 23:50:07 us-atl frps[1384]: 2021/07/29 23:50:07 [D] [vhost.go:147] get hostname from http/https request error: readHandshake: extensionsLen[1375] is not equal to dataLen[910]
Jul 29 23:50:07 us-atl frps[1384]: 2021/07/29 23:50:07 [D] [vhost.go:147] get hostname from http/https request error: readHandshake: extensionsLen[1377] is not equal to dataLen[908]
Jul 29 23:50:08 us-atl frps[1384]: 2021/07/29 23:50:08 [D] [vhost.go:147] get hostname from http/https request error: readHandshake: extensionsLen[1375] is not equal to dataLen[910]
Jul 29 23:50:08 us-atl frps[1384]: 2021/07/29 23:50:08 [D] [vhost.go:147] get hostname from http/https request error: readHandshake: extensionsLen[1377] is not equal to dataLen[908]
Jul 29 23:50:09 us-atl frps[1384]: 2021/07/29 23:50:09 [D] [vhost.go:147] get hostname from http/https request error: readHandshake: extensionsLen[1375] is not equal to dataLen[910]
Jul 29 23:50:09 us-atl frps[1384]: 2021/07/29 23:50:09 [D] [vhost.go:147] get hostname from http/https request error: readHandshake: extensionsLen[1377] is not equal to dataLen[908]
Jul 29 23:50:00 us-atl frps[1384]: 2021/07/29 23:50:00 [I] [proxy.go:162] [] [bitwarden] get a user connection [:48712]
```

**Supplementary information**

**Can you guess what caused this issue**

The parser code for TLS clientHello messages in FRP.

**Checklist**:

- [X] I included all information required in the sections above
- [X] I made sure there are no duplicates of this report [(Use Search)](https://github.com/fatedier/frp/issues?q=is%3Aissue)

**Comment by fatedier** [comment]
Can't reproduce in my local environment.

**Comment by AlainLam** [comment]
I can recreate this, version 0.37.0 and 0.34.3.
It seems that this issue occurs on edge and chrome, and works well on Firefox and IE.

The error message same this:
get hostname from http/https request error: readHandshake: extensionsLen[1377] is not equal to dataLen[908].

I've always wondered if it was my config error, but it works fine on Firefox.

**Comment by mao13820** [comment]
Me too, same problem, my version is 0.21. 

When I used FRP for HTTPS reverse proxy, certain domains cannot be accessed by HTTPS with **chrome 92.0.4515.107** for windows, throwing out ERROR_CONNECTION_RESET error message. 

I have also tested the **newest chrome 93.0.4577.15 beta** for android and got the same problem (I have not test newest chrome for android yet). 

I tested again with **chrome 91.0.4472.57** for windows and **chrome 91.0.4472 for android** but there is no problem for accessing the https website. It seems FRP have compatibility issues with the newest chrome.

I use the newest chrome for windows and take the following test.
Same website, same domain with http method, it  can work well. 
Same website, changing to another domain with https method, it can work well.
Same website, same domain with https method **without FRP**, it can work well.

Very strange problem

**Comment by mao13820** [comment]
I have also tested on edge and ie. It works perfect on IE but failed on edge (version 92.0.902.55)

**Comment by fatedier** [comment]
I tested it on my macos chrome 92 and it success. Does anyone run into this problem using windows?

**Comment by mao13820** [comment]
> I tested it on my macos chrome 92 and it success. Does anyone run into this problem using windows?

I tested it on windows and android, you can see the part I wrote above

**Comment by ProfFan** [comment]
I temporarily downgraded to Chromium 91 and it works. I can see if I can get some packet caps.

**Comment by ProfFan** [comment]
The clientHello that works:
```
Frame 6: 583 bytes on wire (4664 bits), 583 bytes captured (4664 bits) on interface enp0s31f6, id 0
Ethernet II, Src:  (), Dst: : ()
Internet Protocol Version 4, Src: , Dst: 
Transmission Control Protocol, Src Port: 48782, Dst Port: 443, Seq: 1, Ack: 1, Len: 517
Transport Layer Security
    TLSv1.3 Record Layer: Handshake Protocol: Client Hello
        Content Type: Handshake (22)
        Version: TLS 1.0 (0x0301)
        Length: 512
        Handshake Protocol: Client Hello
            Handshake Type: Client Hello (1)
            Length: 508
            Version: TLS 1.2 (0x0303)
            Random: 
            Session ID Length: 32
            Session ID: 
            Cipher Suites Length: 32
            Cipher Suites (16 suites)
            Compression Methods Length: 1
            Compression Methods (1 method)
            Extensions Length: 403
            Extension: Reserved (GREASE) (len=0)
            Extension: server_name (len=16)
            Extension: extended_master_secret (len=0)
            Extension: renegotiation_info (len=1)
            Extension: supported_groups (len=10)
            Extension: ec_point_formats (len=2)
            Extension: session_ticket (len=0)
            Extension: application_layer_protocol_negotiation (len=14)
            Extension: status_request (len=5)
            Extension: signature_algorithms (len=18)
            Extension: signed_certificate_timestamp (len=0)
            Extension: key_share (len=43)
                Type: key_share (51)
                Length: 43
                Key Share extension
                    Client Key Share Length: 41
                    Key Share Entry: Group: Reserved (GREASE), Key Exchange length: 1
                    Key Share Entry: Group: x25519, Key Exchange length: 32
            Extension: psk_key_exchange_modes (len=2)
            Extension: supported_versions (len=11)
            Extension: compress_certificate (len=3)
            Extension: Unknown type 17513 (len=5)
            Extension: Reserved (GREASE) (len=1)
            Extension: padding (len=121)
            Extension: pre_shared_key (len=75)
```

The one doesn't (Chrome 92):
```
TLSv1.3 Record Layer: Handshake Protocol: Client Hello
    Content Type: Handshake (22)
    Version: TLS 1.0 (0x0301)
    Length: 512
    Handshake Protocol: Client Hello
        Handshake Type: Client Hello (1)
        Length: 508
        Version: TLS 1.2 (0x0303)
        Random: 
        Session ID Length: 32
        Session ID: 
        Cipher Suites Length: 32
        Cipher Suites (16 suites)
        Compression Methods Length: 1
        Compression Methods (1 method)
        Extensions Length: 403
        Extension: Reserved (GREASE) (len=0)
        Extension: server_name (len=16)
        Extension: extended_master_secret (len=0)
        Extension: renegotiation_info (len=1)
        Extension: supported_groups (len=10)
        Extension: ec_point_formats (len=2)
        Extension: session_ticket (len=0)
        Extension: application_layer_protocol_negotiation (len=14)
        Extension: status_request (len=5)
        Extension: signature_algorithms (len=18)
        Extension: signed_certificate_timestamp (len=0)
        Extension: key_share (len=43)
            Type: key_share (51)
            Length: 43
            Key Share extension
                Client Key Share Length: 41
                Key Share Entry: Group: Reserved (GREASE), Key Exchange length: 1
                Key Share Entry: Group: x25519, Key Exchange length: 32
        Extension: psk_key_exchange_modes (len=2)
        Extension: supported_versions (len=11)
        Extension: compress_certificate (len=3)
        Extension: Unknown type 17513 (len=5)
        Extension: Reserved (GREASE) (len=1)
        Extension: padding (len=121)
        Extension: pre_shared_key (len=75)
```

**Comment by ProfFan** [comment]
@fatedier 

Line 56, in `https.go`.

```
data := pool.GetBuf(1024)
```
This is the culprit I think.

BTW the parser seriously looks very unpolished. Maybe a mature parsing library should be used instead.

**Comment by ProfFan** [comment]
Fixed in #2502 @fatedier

**Comment by mao13820** [comment]
> @fatedier
> 
> Line 56, in `https.go`.
> 
> ```
> data := pool.GetBuf(1024)
> ```
> 
> This is the culprit I think.
> 
> BTW the parser seriously looks very unpolished. Maybe a mature parsing library should be used instead.

OMG, that works! Brilliant job! Thank you so much.  The length change of the hello message from 512 to 1488 so it was out of buffer.  I'm not sure I understand this bug right?

**Comment by ProfFan** [comment]
The buffer is too small, and Go is a secure language (pun intended) :)

**Comment by ProfFan** [comment]
@fatedier Could you review and merge #2502 ?

**Comment by fatedier** [comment]
@ProfFan It was fixed in #2504 by refactoring vhost codes for HTTPS. You can validate if it works in your scenario.

**Comment by ProfFan** [comment]
Ping @mao13820 @AlainLam for testing
